bakery/macaroon.go

```diff
@@ -0,0 +1,352 @@
+	Namespace  *checkers.Namespace `json:"ns"`
+}
+
+// Clone returns a copy of the macaroon. It does not
+// clone the macaroon's namespace.
+func (m *Macaroon) Clone() *Macaroon {
+	m1 := *m
+	m1.m = m1.m.Clone()
+	m1.caveatData = make(map[string][]byte)
+	for id, data := range m.caveatData {
+		m1.caveatData[id] = data
+	}
+	return &m1
+}
+
+// MarshalJSON implements json.Marshaler by marshaling
+// the macaroon into the original macaroon format if the
+// version is earlier than Version3.
+func (m *Macaroon) MarshalJSON() ([]byte, error) {
+	if m.version < Version3 {
+		if len(m.caveatData) > 0 {
+			return nil, errgo.Newf("cannot marshal pre-version3 macaroon with external caveat data")
+		}
+		return m.m.MarshalJSON()
+	}
+	caveatData := make(map[string]string)
+	for id, data := range m.caveatData {
+		caveatData[base64.RawURLEncoding.EncodeToString([]byte(id))] = base64.RawURLEncoding.EncodeToString(data)
+	}
+	return json.Marshal(macaroonJSON{
+		Macaroon:   m.m,
+		Version:    m.version,
+		CaveatData: caveatData,
+		Namespace:  m.namespace,
+	})
+}
+
+// UnmarshalJSON implements json.Unmarshaler by unmarshaling in a
+// backwardly compatible way - if provided with a previous macaroon
+// version, it will unmarshal that too.
+func (m *Macaroon) UnmarshalJSON(data []byte) error {
+	// First try with new data format.
+	var m1 macaroonJSON
+	if err := json.Unmarshal(data, &m1); err != nil {
+		// If we get an unmarshal error, we won't be able
+		// to unmarshal into the old format either, as extra fields
+		// are ignored.
+		return errgo.Mask(err)
+	}
+	if m1.Macaroon != nil {
+		// We've got macaroon field - it's the new format.
+		if m1.Version < Version3 || m1.Version > LatestVersion {
+			return errgo.Newf("unexpected bakery macaroon version; got %d want %d", m1.Version, Version3)
+		}
+		if got, want := m1.Macaroon.Version(), MacaroonVersion(m1.Version); got != want {
+			return errgo.Newf("underlying macaroon has inconsistent version; got %d want %d", got, want)
+		}
+		caveatData := make(map[string][]byte)
+		for id64, data64 := range m1.CaveatData {
+			id, err := base64Decode([]byte(id64))
+			if err != nil {
+				return errgo.Notef(err, "cannot decode caveat id")
+			}
+			data, err := base64Decode([]byte(data64))
+			if err != nil {
+				return errgo.Notef(err, "cannot decode caveat")
+			}
+			caveatData[string(id)] = data
+		}
+		m.caveatData = caveatData
+		m.m = m1.Macaroon
+		m.namespace = m1.Namespace
+		// TODO should we allow version > LatestVersion here?
+		m.version = m1.Version
+		return nil
+	}
+	// Try to unmarshal from the original format.
+	var m2 *macaroon.Macaroon
+	if err := json.Unmarshal(data, &m2); err != nil {
+		return errgo.Mask(err)
+	}
+	m.m = m2
+	v, err := bakeryVersion(m2.Version())
+	if err != nil {
+		return errgo.Mask(err)
+	}
+	m.version = v
+	m.namespace = legacyNamespace()
+	return nil
+}
+
+// base64Decode base64-decodes the given data.
+// It accepts both standard padded encoding and unpadded
```

Comment:
The both in this sentence seems (at least to me) to imply that they are the only two choices, I had to do a double take.